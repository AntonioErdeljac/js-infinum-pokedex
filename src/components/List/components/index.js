export { default as Item } from './Item';
export { default as Loader } from './Loader';
export { default as BottomPixel } from './BottomPixel';
export { default as Empty } from './Empty';
