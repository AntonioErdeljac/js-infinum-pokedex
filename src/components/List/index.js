import React from 'react';
import { Container, Row, Col } from 'react-bootstrap';

const List = () => {
  return (
    <Container>
      <Row>
        <Col>Test</Col>
      </Row>
    </Container>
  );
};

export default List;
