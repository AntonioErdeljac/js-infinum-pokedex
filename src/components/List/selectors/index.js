const selectors = {
  items: (state) => state.items.data,
  isLoading: (state) => state.items.isLoading,
  nextQuery: (state) => state.items.nextQuery,
  hasLoaded: (state) => state.items.hasLoaded,
};

export default selectors;
