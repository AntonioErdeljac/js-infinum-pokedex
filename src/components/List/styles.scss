@import '../../scss/bootstrap';

.pd-list {
  margin: map-get($map: $spacers, $key: 5) * 3 0px;
}
