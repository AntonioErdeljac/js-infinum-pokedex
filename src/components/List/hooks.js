import { useCallback, useEffect, useMemo } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';

import selectors from './selectors';

import { getItems, resetItems } from '../../store/actions/items';

export const useList = () => {
  const dispatch = useDispatch();

  const items = useSelector(selectors.items, shallowEqual);
  const isLoading = useSelector(selectors.isLoading, shallowEqual);
  const nextQuery = useSelector(selectors.nextQuery, shallowEqual);
  const hasLoaded = useSelector(selectors.hasLoaded, shallowEqual);

  useEffect(() => {
    dispatch(getItems());

    return () => dispatch(resetItems());
  }, [dispatch]);

  const loadMore = useCallback(() => {
    dispatch(getItems(nextQuery));
  }, [dispatch, nextQuery]);

  const hasMore = useMemo(() => hasLoaded && nextQuery, [hasLoaded, nextQuery]);

  return { items, isLoading, loadMore, hasMore };
};
