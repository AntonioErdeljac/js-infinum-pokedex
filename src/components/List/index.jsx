import cn from 'classnames/bind';
import React from 'react';
import { isEmpty } from 'lodash';
import { Container, Row } from 'react-bootstrap';

import styles from './styles.scss';
import { Item, Loader, LoadMore } from './components';
import { useList } from './hooks';

const cx = cn.bind(styles);

const List = () => {
  const { items, isLoading, loadMore, hasMore } = useList();

  if (isLoading && isEmpty(items)) {
    return <Loader />;
  }

  return (
    <Container className={cx('pd-list')}>
      <Row>
        {items.map((item, index) => (
          <Item key={item.name} name={item.name} index={index} />
        ))}
      </Row>
      <LoadMore onClick={loadMore} disabled={!hasMore || isLoading} />
    </Container>
  );
};

export default List;
