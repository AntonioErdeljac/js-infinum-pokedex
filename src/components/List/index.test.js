import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import List from './index';

const makeProps = ({ items, hasLoaded, isLoading } = {}) => ({
  items: items || [],
  hasLoaded: hasLoaded || false,
  isLoading: isLoading || false,
});

const renderComponent = (props) => {
  return shallow(<List {...props} />);
};

describe('List', () => {
  it('renders correctly according to snapshot', () => {
    const wrapper = renderComponent(makeProps({ items: [], hasLoaded: true, isLoading: false }));

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
