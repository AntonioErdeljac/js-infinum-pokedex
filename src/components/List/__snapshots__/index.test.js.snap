// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly according to snapshot 1`] = `
<Memo(FadeIn)>
  <Container
    className="pd-list"
    fluid={false}
  >
    <Row
      className="mt-3"
      noGutters={false}
    >
      <Memo(Item)
        index={0}
        key="Lorem"
        name="Lorem"
      />
      <Memo(Item)
        index={1}
        key="Ipsum"
        name="Ipsum"
      />
    </Row>
  </Container>
</Memo(FadeIn)>
`;
